server/routers/share-recovery/recoverWallet.ts

```diff
@@ -63,11 +73,41 @@ export const recoverWallet = protectedProcedure
         })
       : null;
 
-    const [challenge, recoveryKeyShare] = await Promise.all([
+    const walletPromise = !input.recoveryBackupShareHash
+      ? ctx.prisma.wallet.findFirst({
+          select: {
+            id: true,
+            status: true,
+            publicKey: true,
+          },
+          where: {
+            id: input.walletId,
+            userId: ctx.user.id,
+          },
+        })
+      : null;
+
+    const [
+      challenge,
+      recoveryKeyShare,
+      wallet,
+    ] = await Promise.all([
       challengePromise,
       recoveryKeySharePromise,
+      walletPromise,
     ]);
 
+    const userWallet = recoveryKeyShare?.wallet || wallet;
+    const walletPublicKey = userWallet?.publicKey;
+
+    // If both `walletId` and `recoveryBackupShareHash` are provided, AND `recoveryBackupSharePublicKey` is
+    // EdDSA, we validate a v2 challenge, otherwise a v1 challenge.
+    //
+    // Note this will temporarily break the previous version, where only `walletId` was required to load a
```

Comment:
However, the recovery flow is rarely used, so we just need to deploy the backend and then the frontend without much delay.